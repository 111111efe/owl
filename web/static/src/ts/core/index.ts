export { QWeb } from "./qweb_vdom";
export { EventBus } from "./event_bus";
export { Component } from "./component";
