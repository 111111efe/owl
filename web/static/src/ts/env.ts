import { WEnv } from "./core/component";
import { QWeb } from "./core/qweb_vdom";
import { Registry } from "./core/registry";
import { idGenerator, memoize } from "./core/utils";
import { TemplateLoader } from "./loaders/templates";
import { actionRegistry } from "./registries";
import { RPC } from "./services/ajax";
import {
  ActionManager,
  ActionWidget,
  IActionManager
} from "./store/action_manager";
import { Ajax, IAjax } from "./store/ajax";
import {
  INotificationManager,
  NotificationManager
} from "./store/notifications";
import { IRouter, Router } from "./store/router";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface Env extends WEnv {
  // services
  actionManager: IActionManager;
  ajax: IAjax;
  notifications: INotificationManager;
  router: IRouter;

  // registries
  actionRegistry: Registry<ActionWidget>;

  // helpers
  rpc: IAjax["rpc"];

  // configuration
  debug: boolean;
  isMobile: boolean;
}

interface Loaders {
  loadTemplates: TemplateLoader;
}

interface Services {
  rpc: RPC;
}

type EnvBuilder = (loaders: Loaders, services: Services) => Promise<Env>;
//------------------------------------------------------------------------------
// Environment
//------------------------------------------------------------------------------

/**
 * init returns the main environment for the application.
 *
 * Note that it does not make much sense (except for tests) to have more than
 * one environment. For example, with two environment, the router code in one
 * environment will probably interfere with the code from the other environment.
 *
 * For this reason, the result of init is memoized: every call to
 * this function will actually return the same environment.
 */
export const makeEnv: EnvBuilder = memoize(_makeEnv);

async function _makeEnv(loaders: Loaders, services: Services): Promise<Env> {
  // services
  const qweb = new QWeb();
  const router = new Router();
  const ajax = new Ajax(services.rpc);
  const actionManager = new ActionManager(actionRegistry, ajax);
  const notifications = new NotificationManager();

  // templates
  const templates = await loaders.loadTemplates();
  qweb.addTemplate("default", "<div/>");
  qweb.loadTemplates(templates);

  const env: Env = {
    // Base widget requirements
    qweb,
    getID: idGenerator(),

    actionManager,
    ajax,
    notifications,
    actionRegistry,
    router,

    rpc: ajax.rpc,

    debug: false,
    isMobile: window.innerWidth <= 768
  };
  return env;
}
