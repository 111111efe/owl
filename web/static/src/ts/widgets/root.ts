import { INotification } from "../core/notifications";
import { Widget } from "./widget";
import { Env } from "../env";
import { MenuInfo, MenuItem, getAppAndAction } from "../misc/menu_helpers";
import { ActionStack } from "../services/action_manager";
import { ActionContainer } from "./action_container";
import { HomeMenu } from "./home_menu";
import { Navbar } from "./navbar";
import { Notification } from "./notification";
import { Query } from "../core/router";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

interface State {
  notifications: INotification[];
  stack: ActionStack;
  inHome: boolean;
  currentApp: MenuItem | null;
}

interface Props {
  menuInfo: MenuInfo;
}

//------------------------------------------------------------------------------
// Root Widget
//------------------------------------------------------------------------------

export class Root extends Widget<Props, State> {
  template = "web.web_client";
  widgets = { Navbar, Notification, HomeMenu, ActionContainer };

  state: State = {
    notifications: [],
    stack: [],
    inHome: false,
    currentApp: null
  };

  constructor(env: Env, props: Props) {
    super(env, props);
    const query = this.env.router.getQuery();
    let { app, actionId } = getAppAndAction(props.menuInfo, query);
    this.state.currentApp = app;
    if (!actionId) {
      this.state.inHome = true;
    }
  }
  mounted() {
    // notifications
    this.env.notifications.on("notifications_updated", this, notifs =>
      this.updateState({ notifications: notifs })
    );

    // actions
    this.env.actionManager.on("action_stack_updated", this, stack =>
      this.updateState({ stack, inHome: false })
    );
    this.env.router.on("query_changed", this, this.updateAction);
    this.updateAction(this.env.router.getQuery());
  }

  private updateAction(query: Query) {
    let { app, actionId } = getAppAndAction(this.props.menuInfo, query);
    this.updateAppState(app, actionId);
  }

  private updateAppState(app: MenuItem | null, actionId: number | null) {
    const newApp = app || this.state.currentApp;
    if (actionId) {
      const query: Query = { action_id: String(actionId) };
      const menuId = newApp ? newApp.menuId : false;
      if (menuId) {
        query.menu_id = String(menuId);
      }
      if (app) {
        this.updateState({ currentApp: app });
      }
      this.env.router.navigate(query);
      this.env.actionManager.doAction(actionId);
    } else {
      this.updateState({ inHome: true, currentApp: newApp });
    }
  }

  toggleHome() {
    this.updateState({ inHome: !this.state.inHome });
  }

  openMenu(menu: MenuItem) {
    const app = this.props.menuInfo.menuMap[menu.menuId]!;
    this.updateAppState(app, menu.actionId);
  }
}
