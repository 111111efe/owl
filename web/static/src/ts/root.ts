import { Widget } from "./core/widget";
import { Navbar } from "./widgets/navbar";
import { ActionWidget } from "./services/action_manager";
import { INotification } from "./services/notifications";
import { Notification } from "./widgets/notification";
import { Env } from "./env";

const template = `
    <div class="o_web_client">
        <t t-widget="Navbar"/>
        <div class="o_content" t-ref="content"/>
        <div class="o_notification_container">
          <t t-foreach="state.notifications" t-as="notif">
            <t t-widget="Notification" t-props="notif"/>
          </t>
        </div>
    </div>
`;

export class Root extends Widget<Env, {}> {
  name = "root";
  template = template;
  widgets = { Navbar, Notification };
  content: Widget<Env, {}> | null = null;

  state: { notifications: INotification[] } = { notifications: [] };

  mounted() {
    this.env.actionManager.on("action_ready", this, this.setContentWidget);
    this.env.notifications.on("notification_added", this, this.addNotif);
    this.env.notifications.on("notification_removed", this, this.removeNotif);

    const actionWidget = this.env.actionManager.getCurrentAction();
    if (actionWidget) {
      this.setContentWidget(actionWidget);
    }
  }

  async setContentWidget(actionWidget: ActionWidget) {
    const currentWidget = this.content;
    const newWidget = new actionWidget.Widget(this, actionWidget.props);
    await newWidget.mount(<HTMLElement>this.refs.content);
    if (currentWidget) {
      currentWidget.destroy();
    }
    this.content = newWidget;
  }

  addNotif(notif: INotification) {
    const notifications = this.state.notifications.concat(notif);
    this.updateState({ notifications });
  }
  removeNotif(notif: INotification) {
    const notifs = this.state.notifications.filter(f => f.id !== notif.id);
    this.updateState({ notifications: notifs });
  }
}
