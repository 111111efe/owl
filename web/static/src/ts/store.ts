import { Type } from "./core/component";
import { EventBus } from "./core/event_bus";
import { Registry } from "./core/registry";
import { RPC } from "./services/ajax";
import { IRouter, Query } from "./services/router";
import { Widget } from "./widgets/widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export interface MenuItem {
  id: number;
  name: string;
  parentId: number | false;
  action: string | false;
  icon: string | false;

  // root menu id
  app: MenuItem;
  actionId: number;
  children: MenuItem[];
}

export interface MenuInfo {
  menus: { [key: number]: MenuItem | undefined };

  actionMap: { [id: number]: MenuItem | undefined };
  roots: number[];
}

export interface State {
  stack: ActionStack;
  inHome: boolean;
  currentApp: MenuItem | null;
}

export interface Services {
  rpc: RPC;
  router: IRouter;
}

export type Context = { [key: string]: any };

export interface CommonActionInfo {
  id: number;
  context: Context;
  title: string;
  target: "current" | "new";
}

export type ActionRequest = string | number;

export type ActionWidget = Type<Widget<{}, {}>>;

export interface ClientActionInfo extends CommonActionInfo {
  type: "client";
  name: string;
  Widget: ActionWidget;
}

export interface ActWindowInfo extends CommonActionInfo {
  type: "act_window";
  view: string;
}

export interface ActionDescription {
  id: number;
  type: "ir.actions.act_window" | "ir.actions.client";
  target: "current";
}

export type ActionInfo = ClientActionInfo | ActWindowInfo;
export type ActionStack = ActionInfo[];

export interface RPCModelQuery {
  model: string;
  method: string;
  args?: any[];
  kwargs?: { [key: string]: any };
  context?: { [key: string]: any };
}

export interface RPCControllerQuery {
  route: string;
  params: { [key: string]: any };
}

export type RPCQuery = RPCModelQuery | RPCControllerQuery;

export type RPC = (rpc: RPCQuery) => Promise<any>;

export interface INotification {
  id: number;
  title: string;
  message: string;
  type: "notification" | "warning";
  sticky: boolean;
}

interface RequestParameters {
  route: string;
  params: { [key: string]: any };
}

//------------------------------------------------------------------------------
// Store
//------------------------------------------------------------------------------

export class Store extends EventBus {
  state: State;
  menuInfo: MenuInfo;
  services: Services;
  actionRegistry: Registry<ActionWidget>;

  constructor(
    services: Services,
    menuInfo: MenuInfo,
    actionRegistry: Registry<ActionWidget>
  ) {
    super();
    this.services = services;
    this.menuInfo = menuInfo;
    this.actionRegistry = actionRegistry;
    this.state = {
      stack: [],
      inHome: false,
      currentApp: null
    };
    const query = this.services.router.getQuery();
    let { app, actionId } = this.getAppAndAction(query);
    this.state.currentApp = app;
    if (!actionId) {
      this.state.inHome = true;
    }

    this.services.router.on("query_changed", this, this.updateAction);
    this.updateAction(this.services.router.getQuery());
  }

  private updateAction(query: Query) {
    let { app, actionId } = this.getAppAndAction(query);
    this.updateAppState(app, actionId);
  }

  dispatch(action: string, params?: any) {
    switch (action) {
      case "open_menu":
        this.updateAppState(params.app, params.actionId);
        break;
      case "toggle_home_menu":
        this.updateState({ inHome: !this.state.inHome });
        break;
      case "add_notification":
        this.addNotification(params);
        break;
      case "close_notification":
        this.closeNotification(params);
        break;
    }
  }

  updateState(nextState: Partial<State>) {
    Object.assign(this.state, nextState);
    this.trigger("state_updated", this.state);
  }

  nextID = 1;

  addNotification(notif: Partial<INotification>): number {
    const id = this.nextID++;
    const defaultVals = {
      title: "",
      message: "",
      type: "notification",
      sticky: false
    };
    const notification = Object.assign(defaultVals, notif, { id });
    this.trigger("notification_added", notification);
    if (!notification.sticky) {
      setTimeout(() => this.closeNotification(id), 2500);
    }
    return id;
  }
  closeNotification(id: number) {
    this.trigger("notification_closed", id);
  }

  private updateAppState(app: MenuItem | null, actionId: number | null) {
    const newApp = app || this.state.currentApp;
    if (actionId) {
      const query: Query = { action_id: String(actionId) };
      const menuId = newApp ? newApp.app.id : false;
      if (menuId) {
        query.menu_id = String(menuId);
      }
      if (app) {
        this.updateState({ currentApp: app });
      }
      this.services.router.navigate(query);
      this.doAction(actionId);
    } else {
      this.updateState({ inHome: true, currentApp: newApp });
    }
  }

  private getAppAndAction(
    query: Query
  ): { app: MenuItem | null; actionId: number | null } {
    const menuInfo = this.menuInfo;
    let app: MenuItem | null = null;
    let actionId: number | null = null;
    if ("action_id" in query) {
      actionId = parseInt(query.action_id, 10);
      if (menuInfo.actionMap[actionId]) {
        const menu = menuInfo.actionMap[actionId]!;
        app = menu.app;
      }
    }
    if ("menu_id" in query) {
      const menuId = parseInt(query.menu_id, 10);
      const menu = menuInfo.menus[menuId];
      if (menu) {
        app = menu.app;
        if (!actionId) {
          actionId = menu.actionId;
        }
      }
    }
    return { app, actionId };
  }

  doAction(request: ActionRequest) {
    if (typeof request === "number") {
      this.loadAction(request);
      // this is an action ID
      let name = request === 131 ? "discuss" : "crm";
      let title =
        request === 131 ? "Discuss" : request === 250 ? "Notes" : "CRM";
      let Widget = this.actionRegistry.get(name);
      this.updateState({
        inHome: false,
        stack: [
          {
            id: 1,
            context: {},
            target: "current",
            type: "client",
            name,
            title,
            Widget: Widget
          }
        ]
      });
    }
  }

  private loadAction(id: number) {
    this.rpc({
      route: "web/action/load",
      params: {
        action_id: id
      }
    });
  }

  counter: number = 0;

  async rpc(rpc: RPCQuery): Promise<any> {
    const request = this.prepareRequest(rpc);
    if (this.counter === 0) {
      this.trigger("rpc_status", "loading");
    }
    this.counter++;
    const result = await this.services.rpc(request.route, request.params);
    this.counter--;
    if (this.counter === 0) {
      this.trigger("rpc_status", "notloading");
    }
    return result;
  }

  private prepareRequest(query: RPCQuery): RequestParameters {
    let route: string;
    let params = "params" in query ? query.params : {};
    if ("route" in query) {
      route = query.route;
    } else if ("model" in query && "method" in query) {
      route = `/web/dataset/call_kw/${query.model}/${query.method}`;
      params.args = query.args || [];
      params.model = query.model;
      params.method = query.method;
      params.kwargs = Object.assign(params.kwargs || {}, query.kwargs);
      params.kwargs.context =
        query.context || params.context || params.kwargs.context;
    } else {
      throw new Error("Invalid Query");
    }

    // doing this remove empty keys, and undefined stuff
    const sanitizedParams = JSON.parse(JSON.stringify(params));
    return { route, params: sanitizedParams };
  }
}
