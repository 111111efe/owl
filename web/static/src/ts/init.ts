import { Ajax } from "./store/ajax";
import { NotificationManager } from "./core/notifications";
import { QWeb } from "./core/qweb_vdom";
import { Router } from "./core/router";
import { findInTree, idGenerator, memoize } from "./core/utils";
import { actionRegistry } from "./registries";
import { ActionManager } from "./store/action_manager";
import { MenuInfo, MenuItem } from "./widgets/root";
import { Env } from "./widgets/widget";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

interface InitializedData {
  env: Env;
  menuInfo: MenuInfo;
}

/**
 * init returns the main environment for the application.
 *
 * Note that it does not make much sense (except for tests) to have more than
 * one environment. For example, with two environment, the router code in one
 * environment will probably interfere with the code from the other environment.
 *
 * For this reason, the result of init is memoized: every call to
 * this function will actually return the same environment.
 */
export const init = memoize(async function(): Promise<InitializedData> {
  // services
  const qweb = new QWeb();
  const router = new Router();
  const ajax = new Ajax(ajaxFetch);
  const actionManager = new ActionManager(actionRegistry, ajax);
  const notifications = new NotificationManager();

  // templates
  const templates = await loadTemplates();
  qweb.addTemplate("default", "<div/>");
  qweb.loadTemplates(templates);

  const menuInfo = loadMenus();

  const env: Env = {
    // Base widget requirements
    qweb,
    getID: idGenerator(),

    actionManager,
    ajax,
    notifications,
    actionRegistry,
    router,

    rpc: ajax.rpc,

    debug: false,
    isMobile: window.innerWidth <= 768
  };
  return { env, menuInfo };
});

//------------------------------------------------------------------------------
// Adapters
//------------------------------------------------------------------------------

function ajaxFetch(route: string, params: any): Promise<any> {
  console.log("RPC", route, params);
  const delay = Math.random() * 150;
  return new Promise(resolve => setTimeout(resolve, delay));
}

/**
 * Load xml templates as a string.
 */
async function loadTemplates(): Promise<string> {
  const result = await fetch("templates.xml");
  if (!result.ok) {
    throw new Error("Error while fetching xml templates");
  }
  return result.text();
}

/**
 * Load all menu items
 */
function loadMenus(): MenuInfo {
  const menuItems: BaseMenuItem[] = (<any>window).odoo.menus;
  if (!menuItems) {
    throw new Error("Cannot find menus description");
  }
  const menuInfo = getMenuInfo(menuItems);
  delete (<any>window).odoo.menus; // overkill?
  return menuInfo;
}

interface BaseMenuItem {
  id: number;
  name: string;
  parent_id: number | false;
  action: string | false;
  icon: string | false;
  children: BaseMenuItem[];
}

/**
 * Generate a valid MenuInfo object from a list of BaseMenuItems. This function
 * is supposed to be called once at startup.
 */
export function getMenuInfo(items: BaseMenuItem[]): MenuInfo {
  const menus: { [key: number]: MenuItem | undefined } = {};
  const actionMap: { [id: number]: MenuItem | undefined } = {};
  const roots: number[] = [];

  // build MenuItems
  for (let root of items) {
    roots.push(root.id);
    addToMap(root);
  }

  function addToMap(m: BaseMenuItem, app?: MenuItem): MenuItem {
    const item: Partial<MenuItem> = {
      id: m.id,
      name: m.name,
      parentId: m.parent_id,
      action: m.action,
      icon: m.icon,
      actionId: -1 // will be filled later on with correct id
    };
    item.app = app || (item as MenuItem);
    item.children = m.children.map(c => addToMap(c, item.app as MenuItem));
    menus[item.id!] = item as MenuItem;
    return item as MenuItem;
  }

  // add proper actionId to every menuitems
  for (let menuId in menus) {
    const menu = menus[menuId]!;
    let menuWithAction = menu && findInTree(menu, m => Boolean(m.action));
    if (menuWithAction) {
      menu.actionId = parseInt(
        (<string>menuWithAction.action).split(",")[1],
        10
      );
      if (menuWithAction === menu) {
        actionMap[menu.actionId] = menu;
      }
    }
  }

  return { menus, roots, actionMap };
}
