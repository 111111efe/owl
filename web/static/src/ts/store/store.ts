import { EventBus } from "../core/event_bus";
import { Registry } from "../core/registry";
import { RPC } from "../services/ajax";
import { IRouter, Query } from "../services/router";
import {
  actionManagerMixin,
  ActionStack,
  ActionWidget
} from "./action_manager_mixin";
import { notificationMixin } from "./notification_mixin";
import { rpcMixin } from "./rpc_mixin";
import { MenuItem } from "./store";

//------------------------------------------------------------------------------
// Types
//------------------------------------------------------------------------------

export { ActionStack, ActionWidget } from "./action_manager_mixin";
export { INotification } from "./notification_mixin";
export { RPC } from "./rpc_mixin";

export interface MenuItem {
  id: number;
  name: string;
  parentId: number | false;
  action: string | false;
  icon: string | false;

  // root menu id
  app: MenuItem;
  actionId: number;
  children: MenuItem[];
}

export interface MenuInfo {
  menus: { [key: number]: MenuItem | undefined };

  actionMap: { [id: number]: MenuItem | undefined };
  roots: number[];
}

export interface State {
  stack: ActionStack;
  inHome: boolean;
  currentApp: MenuItem | null;
}

export interface Services {
  rpc: RPC;
  router: IRouter;
}

//------------------------------------------------------------------------------
// Store
//------------------------------------------------------------------------------
export class BaseStore extends EventBus {
  state: State = {
    stack: [],
    inHome: false,
    currentApp: null
  };
  menuInfo: MenuInfo;
  services: Services;
  actionRegistry: Registry<ActionWidget>;
  currentQuery: Query;

  constructor(
    services: Services,
    menuInfo: MenuInfo,
    actionRegistry: Registry<ActionWidget>
  ) {
    super();
    this.services = services;
    this.menuInfo = menuInfo;
    this.actionRegistry = actionRegistry;
    this.currentQuery = {};
  }

  update(nextState: Partial<State>) {
    Object.assign(this.state, nextState);
    this.trigger("state_updated", this.state);
  }

  toggleHomeMenu() {
    if (this.state.inHome && !this.state.currentApp) {
      return;
    }
    this.update({ inHome: !this.state.inHome });
    if (this.state.inHome) {
      this.services.router.navigate({ home: true });
    } else {
      this.updateQuery(this.currentQuery);
    }
  }

  updateQuery(query: Query) {
    this.currentQuery = query;
    this.services.router.navigate(query);
  }
}

export class Store extends actionManagerMixin(
  rpcMixin(notificationMixin(BaseStore))
) {
  constructor(
    services: Services,
    menuInfo: MenuInfo,
    actionRegistry: Registry<ActionWidget>
  ) {
    super(services, menuInfo, actionRegistry);
    const query = this.services.router.getQuery();
    let { app, actionId } = this.getAppAndAction(query);
    this.state.currentApp = app;
    if (!actionId) {
      this.state.inHome = true;
      this.services.router.navigate({ home: true });
    }

    this.services.router.on("query_changed", this, this.updateAction);
    this.updateAction(this.services.router.getQuery());
  }

  private updateAction(query: Query) {
    let { app, actionId } = this.getAppAndAction(query);
    this.updateAppState(app, actionId);
  }

  activateMenuItem(menuId: number) {
    const menu = this.menuInfo.menus[menuId];
    if (!menu) {
      throw new Error("Invalid menu id");
    }
    return this.updateAppState(menu.app, menu.actionId);
  }

  private async updateAppState(app: MenuItem | null, actionId: number | null) {
    const newApp = app || this.state.currentApp;
    if (actionId) {
      const query: Query = { action_id: String(actionId) };
      const menuId = newApp ? newApp.app.id : false;
      if (menuId) {
        query.menu_id = String(menuId);
      }
      if (app) {
        this.update({ currentApp: app });
      }
      await this.doAction(actionId);
      this.updateQuery(query);
    } else {
      this.update({ inHome: true, currentApp: newApp });
    }
  }

  getAppAndAction(
    query: Query
  ): { app: MenuItem | null; actionId: number | null } {
    const menuInfo = this.menuInfo;
    let app: MenuItem | null = null;
    let actionId: number | null = null;
    if (typeof query.action_id === "string") {
      actionId = parseInt(query.action_id, 10);
      if (menuInfo.actionMap[actionId]) {
        const menu = menuInfo.actionMap[actionId]!;
        app = menu.app;
      }
    }
    if (typeof query.menu_id === "string") {
      const menuId = parseInt(query.menu_id, 10);
      const menu = menuInfo.menus[menuId];
      if (menu) {
        app = menu.app;
        if (!actionId) {
          actionId = menu.actionId;
        }
      }
    }
    return { app, actionId };
  }
}
