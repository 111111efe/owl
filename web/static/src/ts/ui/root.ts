import { debounce } from "../core/utils";
import { Env } from "../env";
import { Controller } from "../store/action_manager_mixin";
import { State, Store } from "../store/store";
import { Widget } from "../widget";
import { HomeMenu } from "./home_menu";
import { Navbar } from "./navbar";
import { Notification } from "./notification";

//------------------------------------------------------------------------------
// Root Widget
//------------------------------------------------------------------------------

export class Root extends Widget<Store, State> {
  template = "web.web_client";
  widgets = { Navbar, HomeMenu, Notification };

  store: Store;

  constructor(env: Env, store: Store) {
    super(env, store);
    this.store = store;
    this.state = store.state;
  }

  mounted() {
    this.store.on("state_updated", this, this.updateState);
    this.store.on("rpc_status", this, this.toggleLoadingIndicator);
    this.store.on("update_action", this, this.applyController);
    if (this.store.lastController) {
      this.applyController(this.store.lastController);
    }

    // adding reactiveness to mobile/non mobile
    window.addEventListener("resize", <any>(
      debounce(this.applyMobileInterface.bind(this), 50)
    ));
  }

  applyMobileInterface() {
    const isMobile = window.innerWidth <= 768;
    if (isMobile !== this.env.isMobile) {
      this.env.isMobile = isMobile;
      this.render();
    }
  }

  toggleLoadingIndicator(status: "loading" | "notloading") {
    const method = status === "loading" ? "remove" : "add";
    (<any>this.refs.loading_indicator).classList[method]("d-none");
  }

  async applyController(controller: Controller) {
    const widget = await controller.create(this);
    if (widget) {
      // to do: call some public method of widget instead...
      (<HTMLElement>this.refs.content).appendChild(widget.el!);
      widget.__mount();
      widget.el!.classList.add("o_action_controller");
      this.store.activateController(controller);
    }
  }
}
