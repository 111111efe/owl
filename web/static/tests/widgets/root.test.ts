import { Env, makeEnv } from "../../src/ts/env";
import { Store } from "../../src/ts/store/store";
import { Root } from "../../src/ts/widgets/root";
import * as helpers from "../helpers";

//------------------------------------------------------------------------------
// Setup and helpers
//------------------------------------------------------------------------------

let fixture: HTMLElement;
let store: Store;
let env: Env;
let templates: string;

beforeAll(async () => {
  templates = await helpers.loadTemplates();
});

beforeEach(() => {
  fixture = helpers.makeTestFixture();
  store = helpers.makeTestStore();
  env = makeEnv(store, templates);
  // props = { menuInfo: helpers.makeDemoMenuInfo() };
});

afterEach(() => {
  fixture.remove();
});

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

test("can be rendered (in home menu)", async () => {
  const root = new Root(env, store);
  await root.mount(fixture);
  expect(fixture.innerHTML).toMatchSnapshot();
});

test("if url has action_id, will render action and navigate to proper menu_id", async () => {
  const router = new helpers.MockRouter({ action_id: "595" });
  store = helpers.makeTestStore({ router });
  env = makeEnv(store, templates);
  await helpers.nextTick();

  const root = new Root(env, store);
  await root.mount(fixture);
  expect(env.services.router.getQuery()).toEqual({
    action_id: "595",
    menu_id: "409"
  });
  expect(fixture.innerHTML).toMatchSnapshot();
  // we check here that the url was changed to set app id as menu_id
});

test("start with no action => clicks on client action => discuss is rendered", async () => {
  const root = new Root(env, store);
  await root.mount(fixture);

  expect(env.services.router.getQuery()).toEqual({});

  // discuss menu item
  await (<any>document.querySelector('[data-menu="96"]')).click();
  await helpers.nextTick();
  expect(fixture.innerHTML).toMatchSnapshot();
  expect(env.services.router.getQuery()).toEqual({
    action_id: "131",
    menu_id: "96"
  });
});
