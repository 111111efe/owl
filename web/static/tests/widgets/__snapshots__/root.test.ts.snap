// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be rendered (in home menu) 1`] = `
"<div class=\\"o_web_client\\">
    <div class=\\"o_navbar o_in_home\\">
    
    
  </div>
    
      <div class=\\"o_home_menu\\">
    <div class=\\"o_apps\\">
      <a href=\\"#menu_id=96&amp;action_id=131\\" class=\\"o_app\\" data-menu=\\"96\\">
        <i class=\\"fa fa-comment o_app_icon fa-3x fa-fw\\"></i>
        <div class=\\"o_caption\\">
          Discuss
        </div>
      </a><a href=\\"#menu_id=205&amp;action_id=250\\" class=\\"o_app\\" data-menu=\\"205\\">
        <i class=\\"fa fa-pen o_app_icon fa-3x fa-fw\\"></i>
        <div class=\\"o_caption\\">
          Notes
        </div>
      </a><a href=\\"#menu_id=409&amp;action_id=597\\" class=\\"o_app\\" data-menu=\\"409\\">
        <i class=\\"fa fa-handshake o_app_icon fa-3x fa-fw\\"></i>
        <div class=\\"o_caption\\">
          CRM
        </div>
      </a>
    </div>
  </div>
    
    <div class=\\"o_notification_container\\"></div>
    <div class=\\"o_loading d-none\\">Loading</div>
  </div>"
`;

exports[`if url has action_id, will render action and navigate to proper menu_id 1`] = `
"<div class=\\"o_web_client\\">
    <div class=\\"o_navbar\\">
    <a aria-label=\\"Applications\\" class=\\"o_title fa fa-th\\" href=\\"#\\" title=\\"Applications\\" accesskey=\\"h\\"></a>
    
      <a class=\\"o_menu_brand\\" href=\\"\\" role=\\"button\\">
        CRM
      </a>
      <ul class=\\"o_menu_sections\\">
        <li>
          
            <a href=\\"#\\" role=\\"button\\" class=\\"dropdown-toggle o-no-caret\\" data-toggle=\\"dropdown\\">
              <span>
                Sales
              </span>
            </a>
            <div class=\\"dropdown-menu\\" role=\\"menu\\">
              
                <a href=\\"\\" data-toggle=\\"collapse\\" data-target=\\"#someid\\" role=\\"menuitem\\" class=\\"dropdown-item\\">
                  <span>
                    My Pipeline
                  </span>
                </a>
              
                <a href=\\"\\" data-toggle=\\"collapse\\" data-target=\\"#someid\\" role=\\"menuitem\\" class=\\"dropdown-item\\">
                  <span>
                    My Quotations
                  </span>
                </a>
              
                <a href=\\"\\" data-toggle=\\"collapse\\" data-target=\\"#someid\\" role=\\"menuitem\\" class=\\"dropdown-item\\">
                  <span>
                    Team Pipelines
                  </span>
                </a>
              
            </div>
          
        </li><li>
          
            <a href=\\"#\\" role=\\"button\\" class=\\"dropdown-toggle o-no-caret\\" data-toggle=\\"dropdown\\">
              <span>
                Leads
              </span>
            </a>
            <div class=\\"dropdown-menu\\" role=\\"menu\\">
              
                <a href=\\"\\" data-toggle=\\"collapse\\" data-target=\\"#someid\\" role=\\"menuitem\\" class=\\"dropdown-item\\">
                  <span>
                    Leads
                  </span>
                </a>
              
                <a href=\\"\\" data-toggle=\\"collapse\\" data-target=\\"#someid\\" role=\\"menuitem\\" class=\\"dropdown-item\\">
                  <span>
                    Scoring Rules
                  </span>
                </a>
              
            </div>
          
        </li>
        
      </ul>
    
  </div>
    
      <div class=\\"o_content\\"><div class=\\"o_crm\\">
    <span>CRM!!!!</span>
  </div></div>
    
    <div class=\\"o_notification_container\\"></div>
    <div class=\\"o_loading d-none\\">Loading</div>
  </div>"
`;

exports[`start with no action => clicks on client action => discuss is rendered 1`] = `
"<div class=\\"o_web_client\\">
    <div class=\\"o_navbar\\">
    <a aria-label=\\"Applications\\" class=\\"o_title fa fa-th\\" href=\\"#\\" title=\\"Applications\\" accesskey=\\"h\\"></a>
    
      <a class=\\"o_menu_brand\\" href=\\"\\" role=\\"button\\">
        Discuss
      </a>
      <ul class=\\"o_menu_sections\\">
        <li>
          
            <a href=\\"#\\" role=\\"button\\" class=\\"dropdown-toggle o-no-caret\\" data-toggle=\\"dropdown\\">
              <span>
                Integrations
              </span>
            </a>
            <div class=\\"dropdown-menu\\" role=\\"menu\\">
              
                <a href=\\"\\" data-toggle=\\"collapse\\" data-target=\\"#someid\\" role=\\"menuitem\\" class=\\"dropdown-item\\">
                  <span>
                    Github Repositories
                  </span>
                </a>
              
            </div>
          
        </li>
        
      </ul>
    
  </div>
    
      <div class=\\"o_content\\"><div class=\\"o_discuss\\">
    <span>DISCUSS!!</span>
    <button>Reset first counter</button>
    <button>Reset counter 2 in 3s</button>
    <button>Toggle  Clock/counters</button>
    <button>Toggle Color</button>
    <button>Rerender this widget</button>
    <input>
    
      <div>
      <button>-</button>
      <span style=\\"font-weight:bold\\">Value: 4</span>
      <button>+</button>
    </div>
      <div>
      <button>-</button>
      <span style=\\"font-weight:bold\\">Value: 400</span>
      <button>+</button>
    </div>
    
    <div>
      <span>Current Color: </span>
      red
    </div>
    <button>Add notif</button>
    <button>Add sticky notif</button>
  </div></div>
    
    <div class=\\"o_notification_container\\"></div>
    <div class=\\"o_loading d-none\\">Loading</div>
  </div>"
`;
